smartsim/_core/launcher/dragon/dragonLauncher.py

```diff
@@ -0,0 +1,362 @@
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import fileinput
+import itertools
+import json
+import os
+import subprocess
+import sys
+import typing as t
+from pathlib import Path
+from threading import RLock
+
+import zmq
+
+from ....error import LauncherError
+from ....log import get_logger
+from ....settings import DragonRunSettings, RunSettings, SettingsBase
+from ....status import STATUS_CANCELLED
+from ...config import CONFIG
+from ...schemas import (
+    DragonBootstrapRequest,
+    DragonBootstrapResponse,
+    DragonHandshakeRequest,
+    DragonHandshakeResponse,
+    DragonRequest,
+    DragonRunRequest,
+    DragonRunResponse,
+    DragonStopRequest,
+    DragonStopResponse,
+    DragonUpdateStatusRequest,
+    DragonUpdateStatusResponse,
+)
+from ...utils.network import get_best_interface_and_address
+from ..launcher import WLMLauncher
+from ..step import DragonStep, LocalStep, Step
+from ..stepInfo import StepInfo
+
+logger = get_logger(__name__)
+
+
+class DragonLauncher(WLMLauncher):
+    """This class encapsulates the functionality needed
+    to launch jobs on systems that use Dragon on top of a workload manager.
+
+    All WLM launchers are capable of launching managed and unmanaged
+    jobs. Managed jobs are queried through interaction with with WLM,
+    in this case Dragon. Unmanaged jobs are held in the TaskManager
+    and are managed through references to their launching process ID
+    i.e. a psutil.Popen object
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._context = zmq.Context()
+        self._timeout = CONFIG.dragon_server_timeout
+        self._reconnect_timeout = CONFIG.dragon_server_reconnect_timeout
+        self._startup_timeout = CONFIG.dragon_server_startup_timeout
+        self._context.setsockopt(zmq.SNDTIMEO, value=self._timeout)
+        self._context.setsockopt(zmq.RCVTIMEO, value=self._timeout)
+        self._dragon_head_socket: t.Optional[zmq.Socket[t.Any]] = None
+        self._dragon_head_process: t.Optional[subprocess.Popen[bytes]]
+        self._comm_lock = RLock()
+
+    @property
+    def is_connected(self) -> bool:
+        return self._dragon_head_socket is not None
+
+    def _handsake(self, address: str) -> None:
```

Comment:
Small typo I found while implementing against this PR that I do not want to forget to address:
```suggestion
    def _handshake(self, address: str) -> None:
```

Reply:
LOL